src/policies.rs

```diff
@@ -355,6 +368,73 @@ fn verify_different_authors<G: Git>(
     }
 }
 
+fn verify_rebased<G: Git>(
+    commits: &[Commit],
+    git: &G,
+    old_commit_id: Oid,
+    new_commit_id: Oid,
+    ref_name: &str,
+    override_tag_pattern: &Option<String>,
+) -> Result<PolicyResult, Box<dyn Error>> {
+    let new_branch = old_commit_id.is_zero();
+    let is_merge = git.is_merge_commit(new_commit_id);
+    let is_mainline = git.is_mainline(ref_name)?;
+    let new_commit = git.find_commit(new_commit_id, override_tag_pattern)?;
+    let is_not_fast_forward = !git.is_descendent_of(new_commit_id, old_commit_id)?;
```

Comment:
What if you merged a branch with no common history?

Reply:
Good questions. I'm not sure about the implications of merging in orphan branches. It seems like something this check probably should prevent (orphan branches are many things, but they are almost definitely not rebased on the current release candidate).

This case was added to still support rebasing the RC (for example). I think your 'rebased the entire history' would fall into that camp. It isn't a fast forward push to a merge commit, let it through this check.

Reply:
I checked libgit and it does handle the no common history case correctly. If there's no common history, it returns false.

I think this check is logically correct as is. Something that your question raised for me though is the idea of this being a performance problem if it needs to traverse the whole history.

I set up a pathological test case on Stormsend  (all on my local) because it has a long history. I created an orphan branch with a few hundred commits, and ran these checks. Looks like in that case, is_descendent_of takes about 600ms. I've moved this particular call to is_descendent_of to only run in the case where we actually need the result of the check.

Reply:
Nice catch!